### DISCUSSION TITLE
"PlayBack" mod

### DISCUSSION THREAD

**Discussion by LIII-Works** [ppy/osu#13079] [Ideas]
**Describe the new feature:** 
-
> PlayBack! is a mod that will randomly jump back to part of the timeline when playing a song

I made a video to show you what I mean,sorry I'm bad at editing

**Proposal designs of the feature:** 
-
Please check the video,there is **3** Playback in video

- There should be a small pause after the rewind,sorry I can't show it in the video

**Attached Files**
-

[**_PlayBack!_**](https://streamable.com/3l477)

**Comment by aqilc** [comment]
Hm, I think a better implementation would be that it plays back when you miss so you can practice hard parts in harder maps. In your way, I think it would just be annoying to the users playing with it, because even if you do good on a section, you have to randomly re-play it just because you have bad luck.

I don't think that this should be a mod, but an option in the upcoming(hopefully) practice interface. The idea is pretty great and unique so I think it will make it.

**Comment by LIII-Works** [comment]
Thats true
I define it as a increase difficulty mod

**Comment by aqilc** [comment]
Also, I think that there should be a small pause after the rewind, so the user can understand what's happening and prepare for re-doing the section.

**Comment by LIII-Works** [comment]
> Also, I think that there should be a small pause after the rewind, so the user can understand what's happening and prepare for re-doing the section.

Yes! I want that but I'm unable to show that in video due to my bad editing

**Comment by Kashouryo** [comment]
It shouldn't be categorised as a difficulty increase mod. Because it's based on random instead of player's skill. It's better to put this mod into fun section.

**Comment by LIII-Works** [comment]
> It shouldn't be categorised as a difficulty increase mod. Because it's based on random instead of player's skill. It's better to put this mod into fun section.

That really make sense

**Comment by McKay42** [comment]
I've already had this exact idea implemented (including miss playback and smooth rewind animation + small pause before continuing), but scrapped it due to being unplayable while destroying flow completely.

The gist was that there is no clean way of starting/continuing in the middle of a beatmap, even if you try to add some fancy logic for finding the nearest longest pause between some recent hitobjects.

**Comment by MightyHelper** [comment]
> I don't think that this should be a mod, but an option in the _upcoming(hopefully) practice interface_. The idea is pretty great and unique so I think it will make it.

Is there an Issue/todo list to refer to when mentioning to the practice interface (had no idea but super hyped)

I also suggest that maybe (If ever implemented as a mod)  some system like sd/pf where it could rewind on acc<100%, maybe adjustable via #6272 could be implemented?

**Comment by abraker95** [comment]
> 
> 
> I've already had this exact idea implemented (including miss playback and smooth rewind animation + small pause before continuing), but scrapped it due to being unplayable while destroying flow completely.
> 
> The gist was that there is no clean way of starting/continuing in the middle of a beatmap, even if you try to add some fancy logic for finding the nearest longest pause between some recent hitobjects.

What about going back a bit more and smoothly ramping up speed from 0.0x to 1.0x? The slow ramp up in speed will ensure it is a smooth transition if enough time is given between 0.0x and 1.0x. You will also want the playback be at 1.0x before the target point appears on the screen. Otherwise the pattern displayed from some first note to target note will be subject to the decreased speed as well as differing speed, making target pattern appear different from original play.

So to explain the process:
1. Rewind triggers. The target point in time to stop is `t`
2. Rewind rewinds a bit further than `t` to allow smooth ramp up of playback at some rate.
3. Rewind stops at `t - t_r`, where `t_r` is the extra time needed to ramp up speed.
4. Playback start at x0.0 and increases to x1.0 until `t - t_ar`, which is when target point appears.

```
|---[t - t_r]------------------------[t - t_ar]-------------------[t]------------------------[rewind trigger]---|
       x0.0   x0.25     x0.5    x0.75    x1.0       x1.0     x1.0
```

The only downside in this is that this technically changes the bpm as speed ramps up. I am not sure how weird it will be doing streams and certain other patterns as it ramps up, but I do think this is worth a try and play around with `t_r`.

**Comment by MightyHelper** [comment]
> > I've already had this exact idea implemented (including miss playback and smooth rewind animation + small pause before continuing), but scrapped it due to being unplayable while destroying flow completely.
> > The gist was that there is no clean way of starting/continuing in the middle of a beatmap, even if you try to add some fancy logic for finding the nearest longest pause between some recent hitobjects.
> 
> What about going back a bit more and smoothly ramping up speed from 0.0x to 1.0x? The slow ramp up in speed will ensure it is a smooth transition if enough time is given between 0.0x and 1.0x. You will also want the playback be at 1.0x before the target point appears on the screen. Otherwise the pattern displayed from some first note to target note will be subject to the decreased speed as well as differing speed, making target pattern appear different from original play.
> 
> So to explain the process:
> 
> 1. Rewind triggers. The target point in time to stop is `t`
> 2. Rewind rewinds a bit further than `t` to allow smooth ramp up of playback at some rate.
> 3. Rewind stops at `t - t_r`, where `t_r` is the extra time needed to ramp up speed.
> 4. Playback start at x0.0 and increases to x1.0 until `t - t_ar`, which is when target point appears.
> 
> ```
> |---[t - t_r]------------------------[t - t_ar]-------------------[t]------------------------[rewind trigger]---|
>        x0.0   x0.25     x0.5    x0.75    x1.0       x1.0     x1.0
> ```
> 
> The only downside in this is that this technically changes the bpm as speed ramps up. I am not sure how weird it will be doing streams and certain other patterns as it ramps up, but I do think this is worth a try and play around with `t_r`.

Maybe try to make the time it takes to get from `t - t_r` to `t` equal to some amount of beats ( bpm of the map ) ( maybe 2 or something ), and play the `slider tick sound` at the beats?

Edit:
Even so, I still think that it might be hard to implement, we all have at some time tried to do a 5* with EZ on and the revive is not very comfortable(even though it doesn't rewind).

Idea:
Maybe go back to `t-2t_r` and make the circles from `t-2t_r` to `t-t_r` unhittable, no combo adding and no hp change (basically like holograms to grab the flow back). They could be greyed out in the way that hit circles look like when you die in multi.
